frontend/src/pages/user/shop/ShopSearch.tsx

```diff
@@ -0,0 +1,368 @@
+
+import { Col, Form, Row, Offcanvas } from 'react-bootstrap';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faFilter } from '@fortawesome/free-solid-svg-icons';
+import { Navigate, useNavigate, useParams, useSearchParams } from 'react-router-dom';
+import { SubmitHandler, useForm } from 'react-hook-form';
+import { useState } from 'react';
+import { useQuery } from '@tanstack/react-query';
+
+import GoodsItem from '@components/GoodsItem';
+import TButton from '@components/TButton';
+
+import { PhoneOffCanvasStyle, FilterProps, ProductProps, isDataValid } from '@pages/search';
+import { RouteOnNotOK } from '@lib/Status';
+
+const toNumber = (input: string | null) => {
+  if (!input) return null;
+  const output = Number(input);
+  return isNaN(output) ? null : output;
+};
+
+const toSortBy = (input: string | null) => {
+  switch (input) {
+    case 'price':
+      return 'price';
+    case 'stock':
+      return 'stock';
+    case 'sales':
+      return 'sales';
+    case 'relevancy':
+      return 'relevancy';
+    default:
+      return null;
+  }
+};
+
+const toOrderBy = (input: string | null) => {
+  switch (input) {
+    case 'asc':
+      return 'asc';
+    case 'desc':
+      return 'desc';
+    default:
+      return null;
+  }
+};
+
+const ShopSearch = () => {
+  const navigate = useNavigate();
+  const [searchParams, setSearchParams] = useSearchParams();
+
+  const [showPhoneFilter, setShowPhoneFilter] = useState(false);
+  const phoneFilterOnClick = () => setShowPhoneFilter(!showPhoneFilter);
+
+  const { sellerName } = useParams();
+  const FILTER_OPTIONS: string[] = ['price', 'stock', 'sales', 'relevancy'];
+  const ORDER_OPTIONS: string[] = ['asc', 'desc'];
+
+  const { register, handleSubmit, reset } = useForm<FilterProps>({
+    defaultValues: {
+      minPrice: toNumber(searchParams.get('minPrice')),
+      maxPrice: toNumber(searchParams.get('maxPrice')),
+      minStock: toNumber(searchParams.get('minStock')),
+      maxStock: toNumber(searchParams.get('maxStock')),
+      haveCoupon: Boolean(searchParams.get('haveCoupon')),
+      sortBy: toSortBy(searchParams.get('sortBy')),
+      order: toOrderBy(searchParams.get('order')),
+    },
+  });
+
+  const { data, isError } = useQuery({
+    queryKey: ['shop_search', sellerName, searchParams.toString()],
+    queryFn: async () => {
+      const response = await fetch(`/api/shop/${sellerName}/search?` + searchParams.toString(), {
+        headers: {
+          Accept: 'application/json',
+        },
+      });
+      if (!response.ok) {
+        RouteOnNotOK(response, navigate);
+      }
+      return (await response.json()) as ProductProps[];
+    },
+  });
+
+  const onSubmit: SubmitHandler<FilterProps> = async (data, e) => {
+    e?.preventDefault();
+
+    if (!isDataValid(data)) {
+      return;
+    }
+
+    const params = new URLSearchParams();
+
+    const q = searchParams.get('q') ?? '';
+    if (!q) {
+      return;
+    }
```

Comment:
does this equal to
```js
const q = searchParams.get('q');
```
<details>
<summary>or</summary>

```js
if ( ! searchParams.get('q') ) {
  return;
}
```

</details>
<details>
<summary>or</summary>

```js
const q_is_null = ! searchParams.get('q');
if (q_is_null) {
  return;
}
```

</details>

Reply:
actually non of them above
```js
if ( !searchParams.has('q') ){
  return;
}
const q = searchParams.get('q');
```

Reply:
I believe you mean
```js
const q = searchParams.get('q');
if (!q) {
  return; // no allow q is null
}
params.set('q', q);
```

Reply:
> // no allow q is null

if ur "q" means the variable q, than !q not only block q=null. the statement is incomplete
to clear any ambiguity:
it disallows null or '' from searchParams.get('q')

Reply:
~~I though js player can auto equate null, empty string, 0 and false~~